docs/docs/specs/v0.2/session-graph.md

```diff
@@ -0,0 +1,292 @@
+# `SessionGraph` proposal
+@thure 10 August 2023
+
+# Motivation
+
+We have identified a need for app plugins to be able to:
+
+1. add content items to the UI,
+2. present content items in relation to other content items,
+3. present actions related to a content item,
+4. persist mutations to a content item in the UI,
+
+in such a way that the solution is:
+
+1. independent of any method of persistence beyond the UA’s own JS runtime,
+2. itself able to be persisted or replicated (including sharing),
+3. reasonably extensible by plugins without also causing corruptions.
+
+# Assumptions
+
+This proposal does not assume any particular limits on the host UA’s network bandwidth, memory, disk, or processing.
+
+# Proposed solution
+
+I propose `SessionGraph` should have an ontological structure similar to a relational database, where **Nodes** are stored separately from **Relations**.
+
+## Nodes
+
+Nodes must support the following:
+
+1. A scalar identifier unique within the session
+2. A static or translatable label
+3. A serializable, but possibly arbitrary, record of properties that a plugin can recognize and use as a way to resolve an entity the Node represents
+  - This _may_ include details about provenance
+  - This _may_ include a graphical way to present the node (references to an icon or image)
+
+## Relations
+
+Relations must support the following:
+
+1. One-to-one relation between one Node and another Node
+2. One-to-many relations between one Node and a set of Nodes
+
+## Not stored in `SessionGraph`: functions, class instances, etc
+
+So that the `SessionGraph` can be persisted and replicated, plugins must be able to resolve the entity a Node represents on their own at runtime.
+
+# Implementation
+
+In the rest of this proposal I will refer to the following types:
+
+```tsx
+type Identifier = string;
+
+type NodeKey = `dxos.org/${Identifier}`;
+
+export type Params =
+    | string
+    | number
+    | boolean
+    | null
+    | undefined
+    | Params[]
+    | Record<string, Params>;
+    
+type Node = {
+  id: NodeKey;
+  label: string | [string, {ns: string}];
+  params?: Record<string, Params>;
+};
+
+type RelationType = string;
+
+type RelationKey = `${RelationType}-${ 'child' | 'parent' }`;
```

Comment:
```suggestion
type RelationKey = string;
```

Probably no need to prescribe this, then it can be used to describe other things. (e.g. `create-action` relation to another node which represents the action)

Reply:
Sure, that's fair :)